src/hotspot/share/memory/archiveBuilder.cpp

```diff
@@ -593,6 +595,226 @@ void ArchiveBuilder::make_klasses_shareable() {
+    if (last_obj_end < region_end) {
+      log_debug(cds, map)(PTR_FORMAT ": @@ Misc data " SIZE_FORMAT " bytes",
+                          p2i(last_obj_end + buffer_to_runtime_delta()),
+                          size_t(region_end - last_obj_end));
+      write_data(last_obj_end, region_end, last_obj_end + buffer_to_runtime_delta());
+    }
+  }
+
+#undef _LOG_PREFIX
+
+  // Write information about a region, whose address at dump time is [base .. top). At
+  // runtime, this region will be mapped to runtime_base.  runtime_base is 0 if this
+  // region will be mapped at os-selected addresses (such as the bitmap region), or will
+  // be accessed with os::read (the header).
+  void write_region(const char* name, address base, address top, address runtime_base) {
+    size_t size = top - base;
+    base = runtime_base;
+    top = runtime_base + size;
+    log_info(cds, map)("[%-18s " PTR_FORMAT " - " PTR_FORMAT " " SIZE_FORMAT_W(9) " bytes]",
+                       name, p2i(base), p2i(top), size);
+  }
+
+  // open and closed archive regions
+  void write_heap_region(const char* which, GrowableArray<MemRegion> *regions) {
+    for (int i = 0; i < regions->length(); i++) {
+      address start = address(regions->at(i).start());
+      address end = address(regions->at(i).end());
+      write_region(which, start, end, start);
+      write_data(start, end, start);
+    }
+  }
+
+  // Dump all the data [base...top). Pretend that the base address
+  // will be mapped to runtime_base at run-time.
+  static void write_data(address base, address top, address runtime_base) {
+    assert(top >= base, "must be");
+
+    if (log_is_enabled(Trace, cds, map)) {
+      intx* p = (intx*)base;
+      size_t words = (top - base) / sizeof(uintx);
+
+      for (size_t i = 0; i < words; i += 4, p += 4) {
+        size_t offset = address(p) - base;
+        address runtime_p = runtime_base + offset;
+        switch (words - i) {
+        case 1:
+          log_trace(cds, map)(PTR_FORMAT ": " PTR_FORMAT,
+                              p2i(runtime_p), p[0]);
+          break;
+        case 2:
+          log_trace(cds, map)(PTR_FORMAT ": " PTR_FORMAT " " PTR_FORMAT,
+                              p2i(runtime_p), p[0], p[1]);
+          break;
+        case 3:
+          log_trace(cds, map)(PTR_FORMAT ": " PTR_FORMAT " " PTR_FORMAT " " PTR_FORMAT,
+                              p2i(runtime_p), p[0], p[1], p[2]);
+          break;
+        default:
+          log_trace(cds, map)(PTR_FORMAT ": " PTR_FORMAT " " PTR_FORMAT " " PTR_FORMAT " " PTR_FORMAT,
+                              p2i(runtime_p), p[0], p[1], p[2], p[3]);
+        }
+      }
+    }
+  }
+
+public:
+  CDSMapLogger() {}
+  void write(ArchiveBuilder* builder, FileMapInfo* mapinfo,
+             GrowableArray<MemRegion> *closed_heap_regions,
+             GrowableArray<MemRegion> *open_heap_regions,
+             char* bitmap, size_t bitmap_size_in_bytes) {
```

Comment:
Consider, instead of directly writing to UL, writing to an outputStream*. That would make the code more versatile and reusable. You can then use LogStream to print to UL with the same result.